### PR TITLE
feat: client-side cowswap Tx fees parsing

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Description

This parses fees using market data for Cow Swaps.

They are not expected to be 100% accurate, but it should give us enough confidence to display fees without resorting to using Cow.fi subgraph or parsing Tx events ourselves in unchained.

## Notice

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

- closes https://github.com/shapeshift/web/issues/2279

## Risk

- This should show no regression for non-cow Txs 

## Testing

- CowSwap trades should show fees in sell token under

## Screenshots (if applicable)

image
image

image
image

image
image

### CODE REVIEW

File: src/components/TransactionHistoryRows/utils.ts [hunk 1]
```diff
@@ -58,3 +63,49 @@ export const parseRelevantAssetFromTx = (txDetails: TxDetails, assetType: AssetT
...
+      .div(`1e+${sellAsset.precision}`)
+      .times(bnOrZero(sellAssetPriceAtDate))
+
+    const buyAmountFiat = bnOrZero(buyAmount)
+      .div(`1e+${buyAsset.precision}`)
+      .times(bnOrZero(buyAssetPriceAtDate))
+
+    const sellTokenFee = sellAmountFiat.minus(buyAmountFiat).div(sellAssetPriceAtDate)
+
+    return `${sellTokenFee.toString()} ${sellAsset.symbol}`
```
Comment (gomesalexandre) [inline src/components/TransactionHistoryRows/utils.ts:109]: Self-review: This is wrong, use `<Amount.Crypto />`

Comment (gomesalexandre) [reply src/components/TransactionHistoryRows/utils.ts]: 372465ae09d266f8d8ba3430895d81877c3feff3

File: src/components/TransactionHistoryRows/hooks.tsx [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        txDetails.sellAsset &&
+        txDetails.buyAsset &&
+        txDetails.sellTransfer &&
+        txDetails.buyTransfer
+      )
+    )
+      return
+    if (txDetails.tx.trade.dexName !== Dex.CowSwap) return
+
+    if (!Object.keys(cryptoPriceHistoryData).length) {
```
Comment (gomesalexandre) [inline src/components/TransactionHistoryRows/hooks.tsx:43]: Probably would make more sense to try accessing both AssetIds, which:
1. would be O(1) * 2 vs O(n)
2. might give us more confidence in the fact that these specific AssetIds are loaded

Comment (gomesalexandre) [reply src/components/TransactionHistoryRows/hooks.tsx]: https://github.com/shapeshift/web/pull/2357/commits/396b493807886a7acf0fd945ed621b5f6bd55989

File: src/components/TransactionHistoryRows/utils.ts [hunk 3]
```diff
@@ -58,3 +63,49 @@ export const parseRelevantAssetFromTx = (txDetails: TxDetails, assetType: AssetT
...
+      priceHistoryData: sellAssetPriceHistoryData,
+    })
+
+    const buyAssetPriceAtDate = priceAtDate({
+      date: blockTime,
+      priceHistoryData: buyAssetPriceHistoryData,
+    })
+
+    const sellAmountFiat = bnOrZero(sellAmount)
+      .div(`1e+${sellAsset.precision}`)
```
Comment (gomesalexandre) [inline src/components/TransactionHistoryRows/utils.ts:100]: pow

File: src/components/TransactionHistoryRows/utils.ts [hunk 4]
```diff
@@ -58,3 +63,49 @@ export const parseRelevantAssetFromTx = (txDetails: TxDetails, assetType: AssetT
...
+      .times(bnOrZero(sellAssetPriceAtDate))
+
+    const buyAmountFiat = bnOrZero(buyAmount)
+      .div(`1e+${buyAsset.precision}`)
```
Comment (gomesalexandre) [inline src/components/TransactionHistoryRows/utils.ts:104]: pow

File: src/components/TransactionHistoryRows/hooks.tsx [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+      // Fetch / use cached price history by assetId
+      ;[sellAssetId, buyAssetId].forEach(assetId => {
+        dispatch(findPriceHistoryByAssetId.initiate({ assetId, timeframe: HistoryTimeframe.ALL }))
+      })
+    }
+
+    const tradeFees = getTradeFees({
+      sellAsset,
+      buyAsset,
+      blockTime: dayjs(txDetails.tx.blockTime * 1000).valueOf(),
```
Comment (gomesalexandre) [inline src/components/TransactionHistoryRows/hooks.tsx:60]: comment

File: src/components/TransactionHistoryRows/utils.ts [hunk 6]
```diff
@@ -1,5 +1,10 @@
+import { Asset } from '@shapeshiftoss/asset-service'
 import { cosmos, evm, TransactionMetadata } from '@shapeshiftoss/chain-adapters'
+import { bnOrZero } from '@shapeshiftoss/investor-foxy'
```
Comment (gomesalexandre) [inline src/components/TransactionHistoryRows/utils.ts:3]: tsserver pls

Comment (gomesalexandre) [reply src/components/TransactionHistoryRows/utils.ts]: 323af9e149edcf054e53d0ef552f6e7834d8b95f

File: src/components/TransactionHistoryRows/hooks.tsx (hunk 5 above)
Comment (gomesalexandre) [reply src/components/TransactionHistoryRows/hooks.tsx]: 323af9e149edcf054e53d0ef552f6e7834d8b95f

File: src/components/TransactionHistoryRows/utils.ts (hunk 4 above)
Comment (gomesalexandre) [reply src/components/TransactionHistoryRows/utils.ts]: 323af9e149edcf054e53d0ef552f6e7834d8b95f

File: src/components/TransactionHistoryRows/utils.ts (hunk 3 above)
Comment (gomesalexandre) [reply src/components/TransactionHistoryRows/utils.ts]: 323af9e149edcf054e53d0ef552f6e7834d8b95f

File: src/components/TransactionHistoryRows/utils.ts [hunk 7]
```diff
@@ -58,3 +63,49 @@ export const parseRelevantAssetFromTx = (txDetails: TxDetails, assetType: AssetT
...
+
+    const sellAssetPriceAtDate = priceAtDate({
+      date: blockTime,
+      priceHistoryData: sellAssetPriceHistoryData,
+    })
+
+    const buyAssetPriceAtDate = priceAtDate({
+      date: blockTime,
+      priceHistoryData: buyAssetPriceHistoryData,
+    })
```
Comment (0xdef1cafe) [inline src/components/TransactionHistoryRows/utils.ts:97]: if either of these are `'0'` return null and handle in consumer - it's possible we don't have data here.

**Review by 0xdef1cafe — APPROVED** [review]
"approved" so i can reapprove with changes

**Review by 0xdef1cafe — APPROVED** [review]

**Pull request merged**